src/nvim/window.c

```diff
@@ -5985,3 +6017,18 @@ void win_findbuf(typval_T *argvars, list_T *list)
     }
   }
 }
+
+void win_ui_flush(void)
+{
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if(wp->w_pos_changed) {
+      // make sure grid is allocated before the ui gets any event for it
+      win_grid_alloc(wp, true);
```

Comment:
doing allocation here seems like weird coupling. If we need to make sure the first grid state is sent in the same batch as the first win_position, it would better to delay the sending of `win_position` (by adding `wp->grid.ScreenLines != 0` or something at the if statement here), than to send win_position with a blank grid state in one batch and the real grid state in the next batch.